### PR TITLE
[type] Support bit-level read and write in Python-scope

### PR DESCRIPTION
**Pull request by Hanke98**
Related issue = #1905 #1996 

In this pr, we tried to support reading/writing `bit_struct` in python scope.
The key change is : support `CustomIntType` in `get_data_type`, `get_ret_int` and `get_ret_int`

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: tests/python/test_bit_struct.py [hunk 1]
```diff
@@ -31,6 +31,10 @@ def verify_val():
     set_val()
     verify_val()
 
+    # test bit_struct SNode read and write in python scope
```
Comment (yuanming-hu) [inline tests/python/test_bit_struct.py:34]:
```suggestion
    # Test bit_struct SNode read and write in Python-scope by calling the wrapped, untranslated function body
```

I like this trick :-)

File: tests/python/test_bit_struct.py [hunk 2]
```diff
@@ -66,3 +70,8 @@ def verify_val(idx: ti.i32):
     for idx in range(len(test_case_np)):
         set_val(idx)
         verify_val(idx)
+
+    # test bit_struct SNode read and write in python scope
```
Comment (yuanming-hu) [inline tests/python/test_bit_struct.py:74]:
```suggestion
    # Test bit_struct SNode read and write in Python-scope by calling the wrapped, untranslated function body
```

File: taichi/program/kernel.cpp [hunk 3]
```diff
@@ -295,6 +301,11 @@ int64 Kernel::get_ret_int(int i) {
     return (int64)get_current_program().fetch_result<float32>(i);
   } else if (dt->is_primitive(PrimitiveTypeID::f64)) {
     return (int64)get_current_program().fetch_result<float64>(i);
+  } else if (auto cit = dt->cast<CustomIntType>()) {
+    if (cit->get_is_signed())
+      return (int64)get_current_program().fetch_result<int32>(i);
+    else
+      return (uint64)get_current_program().fetch_result<uint32>(i);
```
Comment (yuanming-hu) [inline taichi/program/kernel.cpp:308]: We are good for now, but sometime in the future we may need to use `i64`.

File: taichi/llvm/llvm_context.cpp [hunk 4]
```diff
@@ -107,6 +107,8 @@ llvm::Type *TaichiLLVMContext::get_data_type(DataType dt) {
     return llvm::Type::getInt32Ty(*ctx);
   } else if (dt->is_primitive(PrimitiveTypeID::u64)) {
     return llvm::Type::getInt64Ty(*ctx);
+  } else if (dt->is<CustomIntType>()) {
+    return llvm::Type::getInt32Ty(*ctx);
```
Comment (yuanming-hu) [inline taichi/llvm/llvm_context.cpp:111]: I don't think this branch belongs here. `CustomIntType` should not be exposed to LLVM. It makes more sense to do the branching on the caller side of this function.

**Review by yuanming-hu — APPROVED** [review]
Thanks! This mostly LGTM. Feel free to merge after the comments are resolved. `TaichiLLVMContext::get_data_type` should not be modified.

**Review by TH3CHARLie — APPROVED** [review]
Looks good and I verified @yuanming-hu's suggestions have been applied.

**Pull request merged**